keras_cv/models/generative/stable_diffusion/diffusion_model.py

```diff
@@ -0,0 +1,279 @@
+            inputs = inputs + [None]
+        x, context = inputs
+        context = x if context is None else context
+        q, k, v = self.to_q(x), self.to_k(context), self.to_v(context)
+        assert len(x.shape) == 3
+        q = tf.reshape(q, (-1, x.shape[1], self.num_heads, self.head_size))
+        k = tf.reshape(k, (-1, context.shape[1], self.num_heads, self.head_size))
+        v = tf.reshape(v, (-1, context.shape[1], self.num_heads, self.head_size))
+
+        q = tf.transpose(q, (0, 2, 1, 3))  # (bs, num_heads, time, head_size)
+        k = tf.transpose(k, (0, 2, 3, 1))  # (bs, num_heads, head_size, time)
+        v = tf.transpose(v, (0, 2, 1, 3))  # (bs, num_heads, time, head_size)
+
+        score = td_dot(q, k) * self.scale
+        weights = keras.activations.softmax(score)  # (bs, num_heads, time, time)
+        attention = td_dot(weights, v)
+        attention = tf.transpose(
+            attention, (0, 2, 1, 3)
+        )  # (bs, time, num_heads, head_size)
+        x = tf.reshape(attention, (-1, x.shape[1], self.num_heads * self.head_size))
+        for layer in self.to_out:
+            x = layer(x)
+        return x
+
+
+class BasicTransformerBlock(keras.layers.Layer):
+    def __init__(self, dim, n_heads, d_head, **kwargs):
+        super().__init__(**kwargs)
+        self.norm1 = keras.layers.LayerNormalization(epsilon=1e-5)
+        self.attn1 = CrossAttention(n_heads, d_head)
+
+        self.norm2 = keras.layers.LayerNormalization(epsilon=1e-5)
+        self.attn2 = CrossAttention(n_heads, d_head)
+
+        self.norm3 = keras.layers.LayerNormalization(epsilon=1e-5)
+        self.geglu = GEGLU(dim * 4)
+        self.dense = keras.layers.Dense(dim)
+
+    def call(self, inputs):
+        inputs, context = inputs
+        x = self.attn1([self.norm1(inputs)]) + inputs
+        x = self.attn2([self.norm2(x), context]) + x
+        return self.dense(self.geglu(self.norm3(x))) + x
+
+
+class Downsample(keras.layers.Layer):
+    def __init__(self, channels, **kwargs):
+        super().__init__(**kwargs)
+        self.conv2d = PaddedConv2D(channels, 3, strides=2, padding=1)
+
+    def call(self, x):
+        return self.conv2d(x)
+
+
+class Upsample(keras.layers.Layer):
+    def __init__(self, channels, **kwargs):
+        super().__init__(**kwargs)
+        self.ups = keras.layers.UpSampling2D(2)
+        self.conv = PaddedConv2D(channels, 3, padding=1)
+
+    def call(self, x):
+        x = self.ups(x)
+        return self.conv(x)
+
+
+class GEGLU(keras.layers.Layer):
+    def __init__(self, output_dim, **kwargs):
+        super().__init__(**kwargs)
+        self.output_dim = output_dim
+        self.dense = keras.layers.Dense(output_dim * 2)
+
+    def call(self, inputs):
+        x = self.dense(inputs)
+        x, gate = x[..., : self.output_dim], x[..., self.output_dim :]
+        return x * gelu(gate)
+
+
+def gelu(x):
+    tanh_res = keras.activations.tanh(x * 0.7978845608 * (1 + 0.044715 * (x**2)))
```

Comment:
numbers seem pretty magical :-)

Reply:
This is just a polynomial approximation of gelu. Generally speaking there are a number of constants harcoded in the code, because the code is non-configurable beyond the arguments of `StableDiffusion` and `text_to_image`. Its only function is to load the original weights and match the original numerics.